HW2/P2/parallel_vector.pyx

```diff
@@ -54,62 +54,98 @@ cpdef move_data_serial(np.int32_t[:] counts,
-
-
+    cdef:
+        int idx, r
+        omp_lock_t *locks = get_N_locks(counts.shape[0])
+
+    ##########
+    # Your code here
+    # Use parallel.prange() and a lock for each element of counts to parallelize
+    # data movement.  Be sure to avoid deadlock, and double-locking.
+    ##########
+    with nogil:
+        for r in prange(repeat, num_threads=4):
+            for idx in range(src.shape[0]):
+                # Always lock on the smaller lock id, this way you avoid deadlocks
+                if src[idx] < dest[idx]:
+                    acquire(&(locks[src[idx]]))
+                    acquire(&(locks[dest[idx]]))
+                # Handle the case where the src and dest lock ids are the same, avoid double locking
+                elif src[idx] == dest[idx]:
+                    acquire(&(locks[src[idx]]))
+                else:
+                    acquire(&(locks[dest[idx]]))
+                    acquire(&(locks[src[idx]]))
+                
+                if counts[src[idx]] > 0:
+                    counts[dest[idx]] += 1
+                    counts[src[idx]] -= 1
+
+                if src[idx] == dest[idx]:
+                    release(&(locks[src[idx]]))
+                else:
+                    release(&(locks[dest[idx]]))
+                    release(&(locks[src[idx]]))
+
+    # Destroy the locks at the end
+    free_N_locks(counts.shape[0], locks)
+
+
+# Move data in parallel with Medium Grained Locks
 cpdef move_data_medium_grained(np.int32_t[:] counts,
                                np.int32_t[:] src,
                                np.int32_t[:] dest,
                                int repeat,
                                int N):
-   cdef:
-       int idx, r
-       int num_locks = (counts.shape[0] + N - 1) / N  # ensure enough locks
-       omp_lock_t *locks = get_N_locks(num_locks)
-
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for every N adjacent elements of counts
-   # to parallelize data movement.  Be sure to avoid deadlock, as well as
-   # double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(num_locks, locks)
+    cdef:
+        int idx, r
+        int num_locks = (counts.shape[0] + N - 1) / N  # ensure enough locks
+        omp_lock_t *locks = get_N_locks(num_locks)
+
+    ##########
+    # Your code here
+    # Use parallel.prange() and a lock for every N adjacent elements of counts
+    # to parallelize data movement.  Be sure to avoid deadlock, as well as
+    # double-locking.
+    ##########
+    with nogil:
+        for r in prange(repeat, num_threads=4):
+            for idx in range(src.shape[0]):
+                # Always lock on the smaller lock id, this way you avoid deadlocks
+                if (src[idx] / N) < (dest[idx] / N):
```

Comment:
Minor things but I'd assign these <var>/N values to a separate variable to help with readability.